File: src/main/scala/queries/VWAP1.scala
package queries

import Math._
import datagen.Bids
import ds.{Cube, Domain, RangeTree, Row, Table}
import ddbt.lib.M3Map
import exec.{DBT, DBT_LMS, Inner, Merge, Naive, ParamsVWAP, VWAPExecutable}
import queries.dbt.VWAP1.TDLLDD
import queries.dbt.VWAP1Base
import utils.{AggPlus, LessThan}
import utils.Helper._

import scala.collection.mutable
import scala.collection.mutable.HashMap

abstract class VWAP1 extends VWAPExecutable {
  def evaluate(bids: Table): (Double, Long)

  override def execute(bids: Table): Long =
    evaluate(bids)._2

  override def query = "Q1"
}

object VWAP1Naive extends VWAP1 {

  import VWAP1._

  override def cost(n: Int, r: Int, p: Int, t: Int): Double = n * n

  override def algo = Naive

  override def evaluate(bids: Table): (Double, Long) = {
    var nC2 = 0.0
    var result = 0.0

    val start = System.nanoTime()
    bids.rows.foreach { b => nC2 += (0.25 * b(volCol)) }
    bids.foreach { b1 => {
      var sum = 0.0
      bids.rows.foreach { b2 =>
        if (b2(priceCol) < b1(priceCol))
          sum += b2(volCol)
      }
      if (nC2 < sum)
        result += b1(priceCol) * b1(volCol)
    }
    }
    val end = System.nanoTime()
    (result, end - start)
  }
}

object VWAP1_DBT_LMS extends VWAP1 {

  import VWAP1._

  override def cost(n: Int, r: Int, p: Int, t: Int): Double = n + p * p

  override def algo = DBT_LMS

  override def evaluate(bids: Table): (Double, Long) = {
    val obj = new VWAP1Base
    //incoming = (p,t,v)
    //outgoing = (t, x, x, v, p)
    val DELTA_BIDS = M3Map.make[TDLLDD, Long]()
    bids.rows.foreach { r => DELTA_BIDS.add(new TDLLDD(r(timeCol), 0, 0, r(volCol), r(priceCol)), 1) }

    val start = System.nanoTime()
    obj.onSystemReady();
    obj.onBatchUpdateBIDS(DELTA_BIDS)
    val end = System.nanoTime()
    (obj.VWAP, end - start)
  }
}

object VWAP1DBT extends VWAP1 {

  import VWAP1._


  override def cost(n: Int, r: Int, p: Int, t: Int): Double = n + p * p

  override def algo = DBT

  override def evaluate(bids: Table): (Double, Long) = {
    val nA = new HashMap[Double, Double]()
    val nC1 = new HashMap[Double, Double]()
    var nC2 = 0.0
    var result = 0.0

    val start = System.nanoTime()
    bids.rows.foreach {
      b =>
        val price = b(priceCol)
        val volume = b(volCol)
        nA += (price -> (nA.getOrElse(price, 0.0) + price * volume))
        nC2 += (0.25 * volume)
        nC1 += (price -> (nC1.getOrElse(price, 0.0) + volume))

    }

    result = 0
    nA.foreach { case (p1, v1) => var sum = 0.0;
      nC1.foreach { case (p2, v2) => if (p2 < p1) sum += v2 };
      if (nC2 < sum) {
        result += v1
      }
    }
    val end = System.nanoTime()
    (result, end - start)
  }


}

object VWAP1Algo1 extends VWAP1 {

  import VWAP1._

  override def cost(n: Int, r: Int, p: Int, t: Int): Double = n + p * log(p)

  override def algo = Inner

  override def evaluate(bids: Table): (Double, Long) = {
    val start = System.nanoTime()
    var result = 0.0

    var nC2 = 0.0
    val nC1 = new mutable.HashMap[Double, Double]()
    bids.rows.foreach { b =>
      val price = b(priceCol)
      val volume = b(volCol)
      nC2 += (0.25 * volume)
      nC1 += (price -> (nC1.getOrElse(price, 0.0) + volume))
    }

    val preAgg = new Table("BidsAgg", nC1.toList.map(kv => Row(Array(kv._1, kv._2))))

    val rtB2 = RangeTree.buildFrom(preAgg, keyVector, 1, valueFn2, AggPlus, "B2")
    val join = rtB2.join(preAgg, keyVector, op.toArray)
    join.foreach { r => if (nC2 < r(aggCol2)) result += r(priceCol) * r(volCol2) }
    val end = System.nanoTime()
    (result, end - start)
  }
}

object VWAP1Algo2 extends VWAP1 {
  override def algo = Merge

  override def cost(n: Int, r: Int, p: Int, t: Int): Double = n + p * log(p)

  override def evaluate(bids: Table): (Double, Long) = {
    import VWAP1._

    val start = System.nanoTime()
    var result = 0.0

    var nC2 = 0.0
    val nC1 = new mutable.HashMap[Double, Double]()
    bids.rows.foreach { b =>
      val price = b(priceCol)
      val volume = b(volCol)
      nC2 += (0.25 * volume)
      nC1 += (price -> (nC1.getOrElse(price, 0.0) + volume))
    }

    val nC1Sorted = nC1.toList.sortBy(_._1)
    val preAgg = new Table("BidsAgg", nC1Sorted.map(kv => Row(Array(kv._1, kv._2))))

    val prices = Domain(nC1Sorted.map(_._1).toArray)
    val cubeB2 = Cube.fromData(Array(prices), preAgg, keyVector, valueFn2, AggPlus)

    cubeB2.accumulate(op)
    val join = cubeB2.join(preAgg, keyVector, op.toArray)

    join.foreach { r => if (nC2 < r(aggCol2)) result += r(priceCol) * r(volCol2) }
    val end = System.nanoTime()
    (result, end - start)
  }
}

object VWAP1 {
  var result = collection.mutable.ListBuffer[Long]()
  var exectime = collection.mutable.ListBuffer[Long]()


  val op = List(LessThan[Double])
  val priceCol = 0
  val timeCol = 1
  val volCol = 2
  val volCol2 = 1
  val aggCol = 3
  val aggCol2 = 2
  val keyVector = (r: Row) => Array(r(priceCol))
  val valueFn = (r: Row) => r(volCol)
  val valueFn2 = (r: Row) => r(volCol2)
  implicit val ord = sorting(keyVector, op)

  val allTests: List[VWAP1] = List(VWAP1Naive, VWAP1_DBT_LMS, VWAP1Algo1, VWAP1Algo2)


  def main(args: Array[String]) = {

    val maxTimeinMS = 1000 * 60 * 60
    (0 to 5).map(13 + 5 * _).foreach { all =>
      var numRuns = 1
      var logn = all
      var logr = all
      var logp = all
      var logt = 10

      if (args.length > 0) {
        logn = args(0).toInt
        logr = args(1).toInt
        logp = args(2).toInt
        logt = args(3).toInt
        numRuns = args(4).toInt
      }

      val bids = new Table("Bids", Bids.loadFromFile(logn, logr, logp, logt))

      allTests.foreach { case a =>
        if (a.enable) {
          exectime.clear();
          (1 to numRuns).foreach { i =>
            val rt = a.evaluate(bids)
            exectime += rt._2
            if (rt._2 / 1000000 > maxTimeinMS)
              a.enable = false
            if (i == numRuns) {
              result += rt._1.toLong
              println(s"${a.query},${a.algo},$logn,$logr,$logp,$logt," + exectime.map(_ / 1000000).mkString(","))
            }
          }

        }
        //println("Res = " + result.mkString(", "))
        //val res = result.head
        //assert(result.map(_ == res).reduce(_ && _))
      }
    }
  }
}

File: srccpp/queries/vwap1.cpp

#include "ds/Table.h"
#include "utils/ComparatorOp.h"
#include "utils/Aggregator.h"
#include "ds/RangeTree.h"
#include "ds/Cube.h"
#include <chrono>
#include <unordered_map>
#include "datagen/BidsGenerator.h"
#include "exec/VWAP.h"
#include "VWAP1_DBT.h"

using namespace std;
using namespace chrono;

struct VWAP1 : BidsExecutable {
    double result;
    static const int priceCol = 0;
    static const int timeCol = 1;
    static const int volCol = 2;
    static const int volCol2 = 1;
    static const int aggCol = 3;
    static const int aggCol2 = 2;
    static const vector<COp> ops;
    static const AggPlus agg;

    VWAP1(const string &algo) : BidsExecutable("Q1", algo) {}

    static void keyFunction(const Row &r, Key &k) {
        k[0] = r[priceCol];
    }

//    static double valueFunction(const Row& r) {
//        return r[volCol];
//    }
    static double valueFunction2(const Row &r) {
        return r[volCol2];
    }

    static const SortingFunc ord;


};

Key k1(1);
Key k2(1);

const vector<COp> VWAP1::ops = {LessThan::getInstance()};
const SortingFunc VWAP1::ord = sorting(&keyFunction, &ops, &k1, &k2);
const AggPlus VWAP1::agg;

struct VWAP1Naive : VWAP1 {
    VWAP1Naive() : VWAP1("Naive,CPP") {}

    long long int evaluate(const Table &bids) override {

        auto start = chrono::steady_clock::now();
        double nC2 = 0;
        result = 0;
        for (const auto &b : bids.rows) {
            nC2 += 0.25 * b[volCol];
        }
        for (const auto &b1: bids.rows) {
            double sum = 0;
            for (const auto &b2: bids.rows) {
                if (b2[priceCol] < b1[priceCol])
                    sum += b2[volCol];
            }
            if (nC2 < sum)
                result += b1[priceCol] * b1[volCol];
        }

        auto end = chrono::steady_clock::now();
        return chrono::duration_cast<chrono::nanoseconds>(end - start).count();
    }
};

struct VWAP1DBT : VWAP1 {

    VWAP1DBT() : VWAP1("DBT,CPP") {}



    long long int evaluate(const Table &bids) override {
        dbtoaster::data_t obj;
//       dbtoaster::data_t::BidsAdaptor::MessageType::
        vector<dbtoaster::BatchMessage<dbtoaster::data_t::BidsAdaptor::MessageType, int>::KVpair> batch(
                bids.rows.size());
        int i = 0;
        for (const auto &r : bids.rows) {
            batch[i].first.content = make_tuple(r[timeCol], 0, 0, r[volCol], r[priceCol]);
            batch[i].second = 1;
            i++;
        }
        auto start = chrono::steady_clock::now();
        obj.on_system_ready_event();
        obj.on_batch_update_BIDS(batch);
        result = obj.get_VWAP();
        auto end = chrono::steady_clock::now();
        return chrono::duration_cast<chrono::nanoseconds>(end - start).count();
    }
};

struct VWAP1Range : VWAP1 {
    VWAP1Range() : VWAP1("Range,CPP") {}

    long long int evaluate(const Table &bids) override {
        auto start = chrono::steady_clock::now();
        result = 0;
        double nC2 = 0;
        unordered_map<double, double> nC1;

        for (const auto &b : bids.rows) {
            int price = b[priceCol];
            int volume = b[volCol];
            nC2 += 0.25 * volume;
            nC1[price] += volume;  //Not correct for max agg;
        }

        Table preAgg;
        for (const auto &kv : nC1) {
            preAgg.rows.emplace_back(Row({kv.first, kv.second}));
        }

        RangeTree rtB2(agg, 1);
        rtB2.buildFrom(preAgg, keyFunction, 1, valueFunction2);
        Table join;
        rtB2.join(preAgg, join, keyFunction, ops);


        for (const auto &r: join.rows) {
            if (nC2 < r[aggCol2])
                result += r[priceCol] * r[volCol2];
        }
        auto end = chrono::steady_clock::now();
        long long ns = chrono::duration_cast<chrono::nanoseconds>(end - start).count();
        return ns;
    }
};

struct VWAP1Merge : VWAP1 {
    VWAP1Merge() : VWAP1("Merge,CPP") {}

    long long int evaluate(const Table &bids) override {
        auto start = chrono::steady_clock::now();
        result = 0;
        double nC2 = 0;
        unordered_map<double, double> nC1;

        for (const auto &b : bids.rows) {
            int price = b[priceCol];
            int volume = b[volCol];
            nC2 += 0.25 * volume;
            nC1[price] += volume;  //Not correct for max agg;
        }

        Table preAgg;
        for (const auto &kv : nC1) {
            preAgg.rows.emplace_back(Row({kv.first, kv.second}));
        }

        sort(preAgg.rows.begin(), preAgg.rows.end(), ord);
        vector<Domain> domains(1);
        preAgg.fillDomain(domains[0], 0, false);

        Cube cubeB2(domains, agg);
        cubeB2.fillData(preAgg, keyFunction, valueFunction2);
        cubeB2.accumulate(ops);

        Table join;
        cubeB2.join(preAgg, join, keyFunction, ops);


        for (const auto &r: join.rows) {
            if (nC2 < r[aggCol2])
                result += r[priceCol] * r[volCol2];
        }
        auto end = chrono::steady_clock::now();
        long long ns = chrono::duration_cast<chrono::nanoseconds>(end - start).count();
        return ns;
    }
};


int main(int argc, char **argv) {


    vector<VWAP1 *> tests;
    tests.emplace_back(new VWAP1Naive);
    tests.emplace_back(new VWAP1DBT);
    tests.emplace_back(new VWAP1Range);
    tests.emplace_back(new VWAP1Merge);
    long long maxTimeInMS = 1000 * 60 * 5;
    int testFlag = 0xFF;
    if (argc >= 3) {
        testFlag = stoi(argv[1]);
        maxTimeInMS = stoi(argv[2]) * 60 * 1000;  //arg in minutes
    }
    bool enable = true;
    for (int all = 10; all <= 28 && enable; all += 1) {
        Table bids;
        int logn = all;
        int logp = all;
        int logt = all;
        int logr = all;
        int numRuns = 1;
        loadFromFile(bids, logn, logr, logp, logt);

        for (int i = 0; i < tests.size(); i++) {
            if (testFlag & (1 << i)) {
                long long execTime = tests[i]->evaluate(bids);
                printf("%s,%s,%d,%d,%d,%d,%lld", tests[i]->query.c_str(), tests[i]->algo.c_str(), logn, logr, logp, logt,
                       execTime / 1000000);
                cout << endl;
                if (execTime / 1000000 > maxTimeInMS)
                    enable = false;
//            cout << "Result = " << (long long) t->result << endl;
            }
        }
    }
}